### PR TITLE
OHRM5X-1001: Employee time sheet with pending action get all API added

### PR DESCRIPTION
**Pull request by VageeshOrangeHRM**
This PR Contains following changes:
1. OHRM5X-1001 - Employee Timesheets - Get Timesheets with Pending action API added.
2. Data group and role permissions added for the endpoint.
3. Unit test covered for the end point.
4. Code refactor added in some files.

### CODE REVIEW

File: installer/dbscript/dbscript-2.sql [hunk 1]
```diff
@@ -4262,6 +4262,7 @@ VALUES ('apiv2_time_time_sheet_config', 'API-v2 Time - Time Sheet Start Day Conf
        ('apiv2_time_project_activity_unique_validation', 'API-v2 Time - Project Activity Unique Validation', 1, 0, 0, 0),
        ('api_v2_time_copy_activities', 'API-v2 Time - Project Copy activities', 1, 1, 0, 0),
        ('apiv2_time_employee_timesheets', 'API-v2 Time - Employee Timesheets', 1, 1, 1, 0),
+       ('apiv2_time_employee_timesheet_action', 'API-v2 Time - Employee Timesheets action', 1, 0, 0, 0),
```
Comment (RajithaKumara) [inline installer/dbscript/dbscript-2.sql:4265]:
```suggestion
       ('apiv2_time_employee_timesheet_list', 'API-v2 Time - Employee Timesheet List', 1, 0, 0, 0),
```

File: installer/dbscript/dbscript-2.sql [hunk 2]
```diff
@@ -4301,6 +4303,7 @@ VALUES ('OrangeHRM\\Time\\Api\\TimeConfigPeriodAPI', @time_module_id, @apiv2_tim
        ('OrangeHRM\\Time\\Api\\TimesheetProjectActivityUniqueValidationAPI', @time_module_id, @apiv2_time_project_activity_unique_validation_data_group_id),
        ('OrangeHRM\\Time\\Api\\CopyProjectActivityAPI', @time_module_id, @api_v2_time_copy_activities_data_group_id),
        ('OrangeHRM\\Time\\Api\\EmployeeTimesheetAPI', @time_module_id, @apiv2_time_employee_timesheets_data_group_id),
+       ('OrangeHRM\\Time\\Api\\EmployeeTimesheetActionAPI', @time_module_id, @apiv2_time_employee_timesheet_action_data_group_id),
```
Comment (RajithaKumara) [inline installer/dbscript/dbscript-2.sql:4306]:
```suggestion
       ('OrangeHRM\\Time\\Api\\EmployeeTimesheetListAPI', @time_module_id, @apiv2_time_employee_timesheet_action_data_group_id),
```

**Review by RajithaKumara — CHANGES_REQUESTED** [review]

File: symfony/plugins/orangehrmTimePlugin/Api/EmployeeTimesheetActionAPI.php [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+        }
+
+        $actions = [WorkflowStateMachine::TIMESHEET_ACTION_APPROVE, WorkflowStateMachine::TIMESHEET_ACTION_REJECT];
+        $actionableStatesList = $this->getUserRoleManager()
+            ->getActionableStates(WorkflowStateMachine::FLOW_TIME_TIMESHEET, $actions);
+        $employeeTimesheetActionSearchParamHolder->setActionableStatesList($actionableStatesList);
+
+        $employeeTimesheetList = $this->getTimesheetService()
+            ->getTimesheetDao()
+            ->getEmployeeTimesheetListByState($employeeTimesheetActionSearchParamHolder);
```
[the remark this answers is not in the document]
Comment (VageeshOrangeHRM) [reply symfony/plugins/orangehrmTimePlugin/Api/EmployeeTimesheetActionAPI.php]: done

File: symfony/plugins/orangehrmTimePlugin/Api/EmployeeTimesheetActionAPI.php [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+
+        $employeeTimesheetListCount = $this->getTimesheetService()
+            ->getTimesheetDao()
+            ->getEmployeeTimesheetListCountByState($employeeTimesheetActionSearchParamHolder);
```
[the remark this answers is not in the document]
Comment (VageeshOrangeHRM) [reply symfony/plugins/orangehrmTimePlugin/Api/EmployeeTimesheetActionAPI.php]: done

File: symfony/plugins/orangehrmTimePlugin/Api/EmployeeTimesheetActionAPI.php [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+use OrangeHRM\Core\Api\V2\Validator\Rule;
+use OrangeHRM\Core\Api\V2\Validator\Rules;
+use OrangeHRM\Core\Traits\UserRoleManagerTrait;
+use OrangeHRM\Entity\Employee;
+use OrangeHRM\Entity\WorkflowStateMachine;
+use OrangeHRM\Time\Api\Model\EmployeeTimesheetModel;
+use OrangeHRM\Time\Dto\EmployeeTimesheetActionSearchFilterParams;
+use OrangeHRM\Time\Traits\Service\TimesheetServiceTrait;
+
+class EmployeeTimesheetActionAPI extends Endpoint implements CollectionEndpoint
```
[the remark this answers is not in the document]
Comment (VageeshOrangeHRM) [reply symfony/plugins/orangehrmTimePlugin/Api/EmployeeTimesheetActionAPI.php]: done

File: symfony/plugins/orangehrmTimePlugin/Dao/TimesheetDao.php [hunk 6]
```diff
@@ -337,7 +338,49 @@ public function isDuplicateTimesheetItem(
...
+    ): Paginator {
+        $q = $this->createQueryBuilder(Timesheet::class, 'timesheet');
+        $q->leftJoin('timesheet.employee', 'employee');
+        $q->andWhere($q->expr()->in('timesheet.state', ':states'));
+
+        if (!is_null($employeeTimesheetActionSearchFilterParams->getEmployeeNumbers())) {
+            $q->andWhere($q->expr()->in('timesheet.employee', ':empNumbers'))
+                ->setParameter('empNumbers', $employeeTimesheetActionSearchFilterParams->getEmployeeNumbers());
+        }
+        $q->setParameter('states', $employeeTimesheetActionSearchFilterParams->getActionableStatesList());
```
[the remark this answers is not in the document]
Comment (VageeshOrangeHRM) [reply symfony/plugins/orangehrmTimePlugin/Dao/TimesheetDao.php]: done

File: symfony/plugins/orangehrmTimePlugin/Dto/EmployeeTimesheetActionSearchFilterParams.php [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+ * You should have received a copy of the GNU General Public License along with this program;
+ * if not, write to the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA  02110-1301, USA
+ */
+
+namespace OrangeHRM\Time\Dto;
+
+use OrangeHRM\Core\Dto\FilterParams;
+
+class EmployeeTimesheetActionSearchFilterParams extends FilterParams
```
[the remark this answers is not in the document]
Comment (VageeshOrangeHRM) [reply symfony/plugins/orangehrmTimePlugin/Dto/EmployeeTimesheetActionSearchFilterParams.php]: done

File: symfony/plugins/orangehrmTimePlugin/config/routes.yaml [hunk 8]
```diff
@@ -167,6 +167,13 @@ apiv2_time_copy_activity:
   defaults:
     _api: OrangeHRM\Time\Api\CopyProjectActivityAPI
 
+apiv2_time_employee_timesheets:
+  path: /api/v2/time/employees/timesheets
```
[the remark this answers is not in the document]
Comment (VageeshOrangeHRM) [reply symfony/plugins/orangehrmTimePlugin/config/routes.yaml]: done

File: symfony/plugins/orangehrmTimePlugin/config/routes.yaml [hunk 9]
```diff
@@ -167,6 +167,13 @@ apiv2_time_copy_activity:
...
+  controller: OrangeHRM\Core\Controller\Rest\V2\GenericRestController::handle
+  methods: [ GET ]
+  defaults:
+    _api: OrangeHRM\Time\Api\EmployeeTimesheetActionAPI
```
[the remark this answers is not in the document]
Comment (VageeshOrangeHRM) [reply symfony/plugins/orangehrmTimePlugin/config/routes.yaml]: done

File: symfony/plugins/orangehrmTimePlugin/test/Api/EmployeeTimesheetActionAPITest.php [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+use OrangeHRM\Tests\Util\EndpointIntegrationTestCase;
+use OrangeHRM\Tests\Util\Integration\TestCaseParams;
+use OrangeHRM\Tests\Util\TestDataService;
+use OrangeHRM\Time\Api\EmployeeTimesheetActionAPI;
+
+/**
+ * @group Time
+ * @group APIv2
+ */
+class EmployeeTimesheetActionAPITest extends EndpointIntegrationTestCase
```
[the remark this answers is not in the document]
Comment (VageeshOrangeHRM) [reply symfony/plugins/orangehrmTimePlugin/test/Api/EmployeeTimesheetActionAPITest.php]: done

File: symfony/plugins/orangehrmTimePlugin/test/Api/EmployeeTimesheetActionAPITest.php [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+        $this->createKernelWithMockServices([Services::AUTH_USER => $this->getMockAuthUser($testCaseParams)]);
+        $this->registerServices($testCaseParams);
+        $this->registerMockDateTimeHelper($testCaseParams);
+        $api = $this->getApiEndpointMock(EmployeeTimesheetActionAPI::class, $testCaseParams);
+        $this->assertValidTestCase($api, 'getAll', $testCaseParams);
+    }
+
+    public function dataProviderForTestGetAll(): array
+    {
+        return $this->getTestCases('EmployeeTimesheetActionTestCase.yaml', 'GetAll');
```
[the remark this answers is not in the document]
Comment (VageeshOrangeHRM) [reply symfony/plugins/orangehrmTimePlugin/test/Api/EmployeeTimesheetActionAPITest.php]: done

File: symfony/plugins/orangehrmTimePlugin/test/Api/EmployeeTimesheetActionAPITest.php [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+ */
+class EmployeeTimesheetActionAPITest extends EndpointIntegrationTestCase
+{
+    /**
+     * @dataProvider dataProviderForTestGetAll
+     */
+    public function testGetAll(TestCaseParams $testCaseParams): void
+    {
+        TestDataService::populate(Config::get(Config::TEST_DIR) . '/phpunit/fixtures/WorkflowStateMachine.yaml');
+        $this->populateFixtures('EmployeeTimesheetActionAPI.yaml', null, true);
```
[the remark this answers is not in the document]
Comment (VageeshOrangeHRM) [reply symfony/plugins/orangehrmTimePlugin/test/Api/EmployeeTimesheetActionAPITest.php]: done

File: symfony/plugins/orangehrmTimePlugin/config/routes.yaml [hunk 13]
```diff
@@ -167,6 +167,13 @@ apiv2_time_copy_activity:
   defaults:
     _api: OrangeHRM\Time\Api\CopyProjectActivityAPI
 
+apiv2_time_employee_timesheets:
```
[the remark this answers is not in the document]
Comment (VageeshOrangeHRM) [reply symfony/plugins/orangehrmTimePlugin/config/routes.yaml]: done

File: installer/dbscript/dbscript-2.sql (hunk 1 above)
Comment (VageeshOrangeHRM) [reply installer/dbscript/dbscript-2.sql]: done

File: installer/dbscript/dbscript-2.sql (hunk 2 above)
Comment (VageeshOrangeHRM) [reply installer/dbscript/dbscript-2.sql]: done

**Review by RajithaKumara — APPROVED** [review]

**Pull request merged**